SCClassLibrary/Common/Audio/UGenResourceManagers.sc

```diff
@@ -0,0 +1,131 @@
+UGenResourceManager {
+	classvar allResourceManagers;
+	*register { |resourceManager| allResourceManagers = allResourceManagers.add(resourceManager) }
+	*createNewInstances { ^allResourceManagers.collect({|m| m -> m.new }).asEvent }
+
+	add { |ugen| this.subclassResponsibility(thisMethod) }
+	panic { |ugen| this.subclassResponsibility(thisMethod) }
+	getAll { this.subclassResponsibility(thisMethod) }
+}
+
+UGenResourceManagerSimple : UGenResourceManager {
+	var previousStateDependentUGen;
+	var all;
+
+	*new { ^super.newCopyArgs(nil, Set.new) }
+	panic { |ugen| this.add(ugen) }
+	getAll { ^all }
+	add { |ugen|
+		previousStateDependentUGen !? { |p|
+			p.createWeakConnectionTo(ugen);
+			all.add(p);
+		};
+		previousStateDependentUGen = ugen;
+	}
+}
+
+UGenResourceManagerWithNonCausalModes : UGenResourceManager {
+	var previousStateDependentUGens;
+	var previousLastGroup;
+	var currentMode;
+	var all;
+
+	var nonCausalModes;
+	var nameOfModeGetter;
+
+	*new { |nonCausalModes, nameOfModeGetter|
+		^super.newCopyArgs(nil, nil, \initial, Set.new, nonCausalModes.asArray, nameOfModeGetter.asSymbol)
+	}
+
+	add { |ugen|
+		if (currentMode == \initial) {
+			currentMode = this.prTryGettingMode(ugen);
+			^previousStateDependentUGens = [ugen];
+		};
+
+		if (currentMode == \panic) {
+			previousStateDependentUGens.do( _.createWeakConnectionTo(ugen) );
+			currentMode = this.prTryGettingMode(ugen);
+			previousStateDependentUGens.do {|p| all.add(p) }
+			^previousStateDependentUGens = [ugen];
+		};
+
+		nonCausalModes.do({ |m|
```

Comment:
```
if(currentMode == m and: {ugen.perform(nameOfModeGetter) == m}){
```
It would be good if this could be factored into a method with a meaningful name to make the code more readable.